### PR TITLE
Implement the Service.Update API call.

### PR DESCRIPTION
**Pull request by frankban**
Also make it the base for other similar service calls,
like changing settings, charm URL or constraints.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4490/
Test PASSed.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 1]
```diff
@@ -1420,75 +1420,103 @@ YUI.add('juju-env-go', function(Y) {
...
-        Type: 'Service',
-        Request: 'SetCharm',
-        Params: {
-          ServiceName: serviceName,
-          CharmUrl: charmUrl,
-          ForceUnits: forceUnits,
-          ForceSeries: forceSeries
+    */
+    setCharm: function(service, url, forceUnits, forceSeries, callback) {
+      var args = {url: url, forceUnits: forceUnits, forceSeries:forceSeries};
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: space between `forceSeries:forceSeries`

File: jujugui/static/gui/src/app/store/env/go.js [hunk 2]
```diff
@@ -2029,16 +2034,15 @@ YUI.add('juju-env-go', function(Y) {
...
-        options) {
+    set_config: function(service, config, callback, options) {
       var ecs = this.get('ecs');
       var args = ecs._getArgs(arguments);
       if (options && options.immediate) {
-        // Need to check that the serviceName is a real service name and not
+        // Need to check that the service is a real service name and not
         // a queued service id before allowing immediate or not.
-        if (ecs.changeSet[serviceName]) {
+        if (ecs.changeSet[service]) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: I always think that `service` is the object. I'd prefer if this stayed its original name `serviceName` as it's more explicit about what it contains.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 3]
```diff
@@ -2029,16 +2034,15 @@ YUI.add('juju-env-go', function(Y) {
       The parameters match the parameters for the public env deploy method in
       go.js.
 
-      @method setConfig
+      @method set_config
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: You could probably add params to this docblock

File: jujugui/static/gui/src/app/store/env/go.js [hunk 4]
```diff
@@ -2049,52 +2053,28 @@ YUI.add('juju-env-go', function(Y) {
...
-       @param {String} serviceName The service name.
-       @param {Object} config The charm configuration options.
-       @param {String} data The YAML representation of the charm
-         configuration options. Only one of `config` and `data` should be
-         provided, though `data` takes precedence if it is given.
-       @param {Object} serviceConfig the current configuration object
-                       of the service.
-       @param {Function} callback A callable that must be called once the
+      @method _set_config
+      @param {String} service The service name.
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: serviceName

File: jujugui/static/gui/src/app/store/env/go.js [hunk 5]
```diff
@@ -2152,57 +2132,6 @@ YUI.add('juju-env-go', function(Y) {
...
-       Change the constraints of the given service.
-
-       @method set_constraints
-       @param {String} serviceName The service name.
-       @param {Object} constraints The new service constraints.
-       @param {Function} callback A callable that must be called once the
-        operation is performed.
-       @return {undefined} Sends a message to the server only.
-    */
-    set_constraints: function(serviceName, constraints, callback) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: We can no longer set the default constraints?

**Comment by hatched** [comment]
:+1: With some trivials 
no qa yet

File: jujugui/static/gui/src/app/store/env/go.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 4 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 5 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: That call wasn't used, and we'll use updateService to set constraints from now on when required.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4494/
Test PASSed.

**Comment by hatched** [comment]
This is looking good! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**